feat(ccp)!: fine grain dataset generation

monoid opened the pull request:
Dataset is generated by fixed size chunks processed by a worker threads. It makes dataset generation cancellable within seconds.

The `[worker]` config section now have a new optional field `dataset_chunk_size` equal by default to 5_000_000.  It is used for interruptable dataset generation, default value corresponds to ca. 10 seconds of uninterruptable work.

Commit:
feat(ccp)!: fine grain dataset generation

Dataset is generated by fixed size chunks processed by a worker threads.
It makes dataset generation cancellable within seconds.

Commit:
Merge branch 'main' into feat/VM-592-fine-grain-dataset

Commit:
Make fmt happy

monoid commented on ccp/src/cu/chunker.rs:
```diff
@@ -0,0 +1,97 @@
+    end: u64,
+}
+
+impl Chunker {
```
ChunkGenerator

monoid commented on ccp/src/cu/chunker.rs:
```diff
@@ -0,0 +1,97 @@
+pub(crate) struct Chunker {
```
Copyrights

Commit:
Review fixes

Commit:
Avoid "chunker" name, except tests

Commit:
Merge branch 'main' into feat/VM-592-fine-grain-dataset

mikevoronov commented on ccp/src/cu/chunk.rs:
```diff
@@ -0,0 +1,113 @@
+}
+
+impl ChunkGenerator {
+    pub(crate) fn new(size: u64, chunk_size: u64) -> Self {
```
Let's call `size` as `dataset_size` or smth like that

mikevoronov commented on ccp/src/cu/chunk.rs:
```diff
@@ -0,0 +1,113 @@
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // we add 1 to substracted and 1 to result because we want to have None when end == start, not Some(0).
+        let remains = self.end.checked_sub(self.start + 1).map(|n| n + 1);
```
I'd suggest to have matching here to early return in case if None to simplify the next line.

mikevoronov replied:
and even better to use if with < or ==, in this case you don't need the logic with +1 and comment to explain it.

mikevoronov commented on ccp/src/cu/chunk.rs:
```diff
@@ -0,0 +1,113 @@
+
+pub(crate) struct ChunkGenerator {
+    // inclusive
+    start: u64,
```
Suggest naming it as `next_chunk_start` b/c it's a bit misleading when you adjust start later

mikevoronov commented on ccp/src/cu/cu_prover.rs:
```diff
@@ -161,22 +173,32 @@ impl CUProver {
-            thread
-                .initialize_dataset(cache.clone(), dataset.clone(), start_item, items_count)
-                .boxed()
+        let chunk_generator = PLMutex::new(ChunkGenerator::new(dataset_size, DATASET_CHUNK_SIZE));
```
I think that you don't need a mutex here, just create separate chackers on equal non intersect intervals.

monoid replied:
Is it guaranteed that chunks of same size need same time for dataset generation?  If not, it will be more efficient to use a shared chunk generator.

Commit:
Review changes

Commit:
Fix tests

Commit:
cu: Use per-thread chunk generators

Commit:
Clean code

mikevoronov commented on ccp/src/cu/cu_prover.rs:
```diff
@@ -161,22 +172,35 @@ impl CUProver {
-                .initialize_dataset(cache.clone(), dataset.clone(), start_item, items_count)
-                .boxed()
+            let chunk_generator =
+                ChunkGenerator::new(start_item, next_start_item, DATASET_CHUNK_SIZE);
```
let's make `DATASET_CHUNK_SIZE` not a constant, but parameter in a config

monoid replied:
Done.

mikevoronov commented on main/src/main.rs:
```diff
@@ -143,11 +144,16 @@ fn build_tokio_runtime(
 async fn async_main(config: CCPConfig, tokio_core_ids_state: CpuIdsHandle) -> eyre::Result<()> {
     let rpc_bind_address = (config.rpc_endpoint.host.clone(), config.rpc_endpoint.port);
     let facade_queue_size = config.rpc_endpoint.facade_queue_size;
+    let global_cancellation_token = CancellationToken::new();
```
Looks like it's not global, but prover cancellation token?

mikevoronov reviewed (approved):
utACK

Commit:
Move dataset chunk size to config

The `[worker]` section now have a new optional field
`dataset_chunk_size` equal by default to 5_000_000.  It is used for
interruptable dataset generation, default value correcponds to ca.
10 seconds of uninterrupted work.
